File: pulsar-common/src/main/java/org/warps/pulsar/common/ObjectCache.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.warps.pulsar.common;

import org.apache.hadoop.conf.Configuration;
import org.warps.pulsar.common.config.ImmutableConfig;

import java.util.HashMap;
import java.util.WeakHashMap;

/**
 * TODO: use BeanFactory(Spring or Apache Configuration)
 * */
public class ObjectCache {

    private static final WeakHashMap<Configuration, ObjectCache> CACHE = new WeakHashMap<>();

    private final HashMap<String, Object> objectMap = new HashMap<>();

    private ObjectCache() {
    }

    public static ObjectCache get(ImmutableConfig conf) {
        return get(conf.unbox());
    }

    public static ObjectCache get(Configuration conf) {
        ObjectCache objectCache = CACHE.get(conf);
        if (objectCache == null) {
            objectCache = new ObjectCache();
            CACHE.put(conf, objectCache);
        }

        return objectCache;
    }

    public boolean hasBean(String key) {
        return objectMap.get(key) != null;
    }

    public Object getBean(String key) {
        return objectMap.get(key);
    }

    @SuppressWarnings("unchecked")
    public <T> T getBean(String key, T defaultValue) {
        Object obj = objectMap.get(key);
        if (obj == null) {
            return defaultValue;
        }
        return (T) obj;
    }

    public <T> T getBean(Class<T> clazz) {
        return (T)objectMap.get(clazz.getName());
    }

    public void put(String key, Object value) {
        objectMap.put(key, value);
    }

    public <T> void put(Object obj) {
        objectMap.put(obj.getClass().getName(), obj);
    }
}

File: pulsar-core/pulsar-common/src/main/kotlin/ai/platon/pulsar/common/ObjectCache.kt
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package ai.platon.pulsar.common

import ai.platon.pulsar.common.config.ImmutableConfig
import java.util.*

/**
 * Object cache, used to cache objects that are expensive to create.
 */
class ObjectCache private constructor() {
    private val objectMap = HashMap<String, Any?>()
    
    fun hasBean(key: String): Boolean {
        return objectMap[key] != null
    }
    
    fun getBean(key: String): Any? {
        return objectMap[key]
    }
    
    fun <T> getBean(key: String, defaultValue: T): T {
        val obj = objectMap[key] ?: return defaultValue
        return obj as T
    }
    
    fun <T> getBean(clazz: Class<T>): T? {
        return objectMap[clazz.name] as T?
    }
    
    inline fun <reified T: Any> getBean(): T? = getBean(T::class.java)

    inline fun <reified T: Any> computeIfAbsent(mappingFunction: () -> T): T {
        val oldValue = getBean<T>()
        val value: T = oldValue ?: mappingFunction()
        if (oldValue == null) {
            putBean(value)
        }
        return value
    }
    
    fun putBean(obj: Any) {
        objectMap[obj.javaClass.name] = obj
    }
    
    companion object {
        private val CACHE = WeakHashMap<ImmutableConfig, ObjectCache>()
        
        @JvmStatic
        fun get(conf: ImmutableConfig): ObjectCache {
            return CACHE.computeIfAbsent(conf) { ObjectCache() }
        }
    }
}
